### PR TITLE
Fetching of ChaosEngine Varibles, and overriding the default variables

### PR DESCRIPTION
**Pull request by rahulchheda**
Signed-off-by: Rahul M Chheda <rahul.chheda@mayadata.io>

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Checklist**

* [x] Does this PR have a corresponding GitHub issue?
* [x] Have you included relevant README for the chaoslib/experiment with details?
* [ ] Have you added debug messages where necessary? 
* [ ] Have you added task comments where necessary? 
* [ ] Have you tested the changes for possible failure conditions?
* [ ] Have you provided the positive & negative test logs for the litmusbook execution?
* [ ] Does the litmusbook ensure idempotency of cluster state?, i.e., is cluster restored to original state?
* [ ] Have you used non-shell/command modules for Kubernetes tasks?
* [ ] Have you (jinja) templatized custom scripts that is run by the litmusbook, if any? 
* [ ] Have you (jinja) templatized Kubernetes deployment manifests used by the litmusbook, if any?
* [ ] Have you reused/created util functions instead of repeating tasks in the litmusbook?
* [ ] Do the artifacts follow the appropriate directory structure? 
* [ ] Have you isolated storage (eg: OpenEBS) specific implementations, checks? 
* [ ] Have you isolated platform (eg: baremetal kubeadm/openshift/aws/gcloud) specific implementations, checks?  
* [ ] Are the ansible facts well defined? Is the scope explicitly set for playbook & included utils?
* [ ] Have you ensured minimum/careful usage of shell utilities (awk, grep, sed, cut, xargs etc.,)?
* [ ] Can the limtusbook be executed both from within & outside a container (configurable paths, no hardcode)?
* [ ] Can you suggest the minimal resource requirements for the litmusbook execution?
* [ ] Does the litmusbook job artifact carry comments/default options/range for the ENV tunables?
* [ ] Has the litmusbooks been linted? 

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by ksatchit** [comment]
@rahulchheda , can you please share the example schema of the chaosengine with components passed & an output log of the executor & chaos job execution with the passed env values.

**Comment by rahulchheda** [comment]
ENGINE SPEC:
apiVersion: litmuschaos.io/v1alpha1
kind: ChaosEngine
metadata:
  name: engine-nginx
  namespace: default
spec:
  monitorenable: true 
  appinfo:
    appkind: deployment
    applabel: app=nginx
    appns: default    
  chaosServiceAccount: nginx
  experiments:
  - name: pod-delete
    spec:
      components:
      - name: PODNAME
        value: nginx
      - name: TARGET_CONTAINER
        value: jack
      - name: RANDOM
        value: kgb
  - name: container-kill
    spec:
      components:
      - name: TARGET_CONTAINER
        value: jackma

**Comment by rahulchheda** [comment]
EXECUTOR VARIABLE FETCHED:
included: /executor/experiment_env_patcher.yml for 127.0.0.1 => (item=[u'APP_NAMESPACE', u'default'])
included: /executor/experiment_env_patcher.yml for 127.0.0.1 => (item=[u'APP_LABEL', u'app=nginx'])
included: /executor/experiment_env_patcher.yml for 127.0.0.1 => (item=[u'APP_KIND', u''])
included: /executor/experiment_env_patcher.yml for 127.0.0.1 => (item=[u'CHAOSENGINE', u'engine-nginx'])
included: /executor/experiment_env_patcher.yml for 127.0.0.1 => (item=[u'ANSIBLE_STDOUT_CALLBACK', u'default'])
included: /executor/experiment_env_patcher.yml for 127.0.0.1 => (item=[u'TOTAL_CHAOS_DURATION', u'15'])
included: /executor/experiment_env_patcher.yml for 127.0.0.1 => (item=[u'CHAOS_INTERVAL', u'5'])
included: /executor/experiment_env_patcher.yml for 127.0.0.1 => (item=[u'LIB', u''])
included: /executor/experiment_env_patcher.yml for 127.0.0.1 => (item=[u'PODNAME', u'nginx'])
included: /executor/experiment_env_patcher.yml for 127.0.0.1 => (item=[u'TARGET_CONTAINER', u'jack'])
included: /executor/experiment_env_patcher.yml for 127.0.0.1 => (item=[u'RANDOM', u'kgb'])

**Comment by rahulchheda** [comment]
POD-DELETE NEW VARIABLES:
apiVersion: batch/v1
kind: Job
metadata:
  creationTimestamp: 2019-09-23T13:37:09Z
  generateName: container-kill-
  labels:
    experiment: container-kill
  name: container-kill-rwdw2
  namespace: default
  resourceVersion: "6690252"
  selfLink: /apis/batch/v1/namespaces/default/jobs/container-kill-rwdw2
  uid: 3dd7e2b1-de07-11e9-8eaa-0050569846e3
spec:
  backoffLimit: 6
  completions: 1
  parallelism: 1
  selector:
    matchLabels:
      controller-uid: 3dd7e2b1-de07-11e9-8eaa-0050569846e3
  template:
    metadata:
      creationTimestamp: null
      labels:
        controller-uid: 3dd7e2b1-de07-11e9-8eaa-0050569846e3
        experiment: container-kill
        job-name: container-kill-rwdw2
    spec:
      containers:
      - args:
        - -c
        - ansible-playbook ./experiments/chaos/kubernetes/container_kill/test.yml
          -i /etc/ansible/hosts -vv; exit 0
        command:
        - /bin/bash
        env:
        - name: ANSIBLE_STDOUT_CALLBACK
          value: default
        - name: APP_NAMESPACE
          value: default
        - name: APP_LABEL
          value: app=nginx
        - name: TARGET_CONTAINER
          value: jackma
        - name: CHAOSENGINE
          value: engine-nginx
        - name: APP_KIND
        - name: KILL_MODE
        - name: LIB
        image: openebs/ansible-runner:ci
        imagePullPolicy: Always
        name: ansibletest
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Never
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: nginx
      serviceAccountName: nginx
      terminationGracePeriodSeconds: 30
status:
  completionTime: 2019-09-23T13:40:13Z
  conditions:
  - lastProbeTime: 2019-09-23T13:40:13Z
    lastTransitionTime: 2019-09-23T13:40:13Z
    status: "True"
    type: Complete
  startTime: 2019-09-23T13:37:09Z
  succeeded: 1

**Comment by rahulchheda** [comment]
CONTAINER-KILL REPLACED VARIABLES:
apiVersion: batch/v1
kind: Job
metadata:
  creationTimestamp: 2019-09-23T13:37:09Z
  generateName: container-kill-
  labels:
    experiment: container-kill
  name: container-kill-rwdw2
  namespace: default
  resourceVersion: "6690252"
  selfLink: /apis/batch/v1/namespaces/default/jobs/container-kill-rwdw2
  uid: 3dd7e2b1-de07-11e9-8eaa-0050569846e3
spec:
  backoffLimit: 6
  completions: 1
  parallelism: 1
  selector:
    matchLabels:
      controller-uid: 3dd7e2b1-de07-11e9-8eaa-0050569846e3
  template:
    metadata:
      creationTimestamp: null
      labels:
        controller-uid: 3dd7e2b1-de07-11e9-8eaa-0050569846e3
        experiment: container-kill
        job-name: container-kill-rwdw2
    spec:
      containers:
      - args:
        - -c
        - ansible-playbook ./experiments/chaos/kubernetes/container_kill/test.yml
          -i /etc/ansible/hosts -vv; exit 0
        command:
        - /bin/bash
        env:
        - name: ANSIBLE_STDOUT_CALLBACK
          value: default
        - name: APP_NAMESPACE
          value: default
        - name: APP_LABEL
          value: app=nginx
        - name: TARGET_CONTAINER
          value: jackma
        - name: CHAOSENGINE
          value: engine-nginx
        - name: APP_KIND
        - name: KILL_MODE
        - name: LIB
        image: openebs/ansible-runner:ci
        imagePullPolicy: Always
        name: ansibletest
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Never
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: nginx
      serviceAccountName: nginx
      terminationGracePeriodSeconds: 30
status:
  completionTime: 2019-09-23T13:40:13Z
  conditions:
  - lastProbeTime: 2019-09-23T13:40:13Z
    lastTransitionTime: 2019-09-23T13:40:13Z
    status: "True"
    type: Complete
  startTime: 2019-09-23T13:37:09Z
  succeeded: 1

File: executor/engine_env_getter.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+- name: Get ENV names for Chaos Engine
+  shell: >
+    kubectl get chaosengine engine-nginx -o json 
+    | jq '.spec.experiments[] | select(.name == "{{ c_experiment_name }}") | .spec.components[{{item}}].name' | tr -d "\""
```
Comment (ksatchit) [inline executor/engine_env_getter.yml]: lets just check if jsonpath filter helps - to try and minimize use of jq

**Comment by ksatchit** [comment]
Can we also introduce an experiment_config_map_parser to get the any configmaps and create them locally? So that executor can include it to construct job spec.

File: executor/engine_env_getter.yml (hunk 1 above)
Comment (ksatchit) [reply executor/engine_env_getter.yml]: @rahulchheda , can we change the engine name to variable

**Comment by ksatchit** [comment]
> Can we also introduce an experiment_config_map_parser to get the any configmaps and create them locally? So that executor can include it to construct job spec.

This is implemented in PR https://github.com/litmuschaos/litmus/pull/794

**Comment by ksatchit** [comment]
@rahulchheda can you add an example to the README saying how experiment values can be overrided, i.e., example engine spec with the component override.

**Pull request merged**